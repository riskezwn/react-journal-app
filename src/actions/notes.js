import { addDoc, collection, doc, updateDoc } from "firebase/firestore";
import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        const newNote = {
            title: "",
            body: "",
            date: new Date().getTime(),
        };
        const collRef = collection(db, `${uid}/journal/notes`);
        const doc = await addDoc(collRef, newNote);
        dispatch(activeNote(doc.id, newNote));
    };
};

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note,
    },
});

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    };
};

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes,
});

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        if (!note.url) {
            delete note.url;
        }

        const noteToFirestore = { ...note };
        delete noteToFirestore.id;

        const docRef = doc(db, `${uid}/journal/notes/${note.id}`);

        try {
            await updateDoc(docRef, noteToFirestore);
            dispatch(refreshNote(note.id, noteToFirestore));
            Swal.fire("Saved!", note.title, "success");
        } catch (e) {
            Swal.fire("Something go wrong", e, "error");
        }
    };
};

export const refreshNote = (id, note) => ({
    type: types.notesUpdate,
    payload: {
        id,
        note: {
            id,
            ...note,
        },
    },
});
