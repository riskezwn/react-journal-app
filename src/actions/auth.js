import {
    createUserWithEmailAndPassword,
    signInWithEmailAndPassword,
    signInWithPopup,
    signOut,
    updateProfile,
} from "firebase/auth";
import Swal from "sweetalert2";
import { auth, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";
import { notesLogout } from "./notes";
import { finishLoading, startLoading } from "./ui";

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());
        signInWithEmailAndPassword(auth, email, password)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading());
            })
            .catch((e) => {
                Swal.fire("Something go wrong :(", "User not found", "error");
                dispatch(finishLoading());
            });
    };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        dispatch(startLoading());
        createUserWithEmailAndPassword(auth, email, password)
            .then(async ({ user }) => {
                await updateProfile(user, { displayName: name });
                dispatch(login(user.uid, user.displayName));
                dispatch(finishLoading());
            })
            .catch((e) => {
                Swal.fire(
                    "Something go wrong :(",
                    "Registration could not be completed",
                    "error"
                );
                dispatch(finishLoading());
            });
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
            dispatch(login(user.uid, user.displayName));
        });
    };
};

export const startLogout = () => {
    return async (dispatch) => {
        await signOut(auth);
        dispatch(logout());
        dispatch(notesLogout());
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName,
    },
});

export const logout = () => ({
    type: types.logout,
});
