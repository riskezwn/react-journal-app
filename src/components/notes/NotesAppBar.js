import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNote } from "../../actions/notes";

export const NotesAppBar = () => {
    const dispatch = useDispatch();
    const { active } = useSelector((state) => state.notes);

    const handleSaveNote = () => {
        dispatch(startSaveNote(active));
    };

    return (
        <div className="notes__appbar">
            <span>28th of August of 2022</span>
            <div>
                <button className="btn">Picture</button>
                <button className="btn" onClick={handleSaveNote}>
                    Save
                </button>
            </div>
        </div>
    );
};
