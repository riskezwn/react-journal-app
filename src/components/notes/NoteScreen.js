import React from "react";
import { NotesAppBar } from "./NotesAppBar";

export const NoteScreen = () => {
    return (
        <div className="notes__main-content">
            <NotesAppBar />
            <div className="notes__content">
                <input
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                />
                <textarea
                    placeholder="What happened today?"
                    className="notes__textarea"
                    autoComplete="off"
                ></textarea>
                <div className="notes__image">
                    <img
                        src="https://i.blogs.es/0ca5da/ambulo_polar_wide/840_560.jpg"
                        alt="imagen"
                    />
                </div>
            </div>
        </div>
    );
};
