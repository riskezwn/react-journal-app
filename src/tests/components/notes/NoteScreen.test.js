/**
 * @jest-environment jsdom
 */

import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { activeNote } from "../../../actions/notes";
import { NoteScreen } from "../../../components/notes/NoteScreen";

import "../../../setupTest";

jest.mock("../../../actions/notes", () => ({
  activeNote: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {
    uid: "JVqvnEtZplUHtVwMZxNQhRDQNdt2",
    name: "riskezwn",
  },
  ui: {
    loading: false,
    msgError: null,
  },
  notes: {
    active: {
      id: 1234,
      title: "hola",
      body: "mundo",
      date: 0,
    },
    notes: [],
  },
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <NoteScreen />
  </Provider>
);

describe("test in <NoteScreen/>", () => {
  test("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should init activeNote", () => {
    wrapper.find('input[name="title"]').simulate("change", {
      target: {
        name: "title",
        value: "hola de nuevo",
      },
    });

    expect(activeNote).toHaveBeenCalledWith(1234, {
      body: "mundo",
      title: "hola de nuevo",
      id: 1234,
      date: 0,
    });
  });
});
