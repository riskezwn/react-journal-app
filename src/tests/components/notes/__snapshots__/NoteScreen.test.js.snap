// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <NoteScreen/> should match snapshot 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      Thursday, 1st of January of 1970
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      autoFocus={true}
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      spellCheck="false"
      type="text"
      value="hola"
    />
    <textarea
      autoComplete="off"
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today?"
      spellCheck="false"
      value="mundo"
    />
  </div>
  <div
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </div>
</div>
`;
